### PR TITLE
Add initial client and server.

### PR DESCRIPTION
**Pull request by MetalMari**

### CODE REVIEW

**Review by mmishche — CHANGES_REQUESTED** [review]

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+make proto:
+	protoc --go_out=. --go_opt=paths=source_relative     --go-grpc_out=. --go-grpc_opt=paths=source_relative     chat_protos/chat.proto
```
[the remark this answers is not in the document]
Comment (MetalMari) [reply Makefile]: Done.

File: client/chatclient/chat_client.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+// GetUsers gets and prints list of users.
+func (c *Client) GetUsers() {
+	ctx := context.Background()
+
+	r, err := c.client.GetUsers(ctx, &pb.GetUsersRequest{})
+	if err != nil {
+		log.Fatalf("could not get users: %v", err)
+	}
+	log.Printf("Users: %s", r.Users)
+}
```
Comment (mmishche) [inline client/chatclient/chat_client.go:44]: I believe the method should return the users and the caller will do anything else
Let's imagine that you will implement a new service that will be using the code above, but the service will filter users by the first letter of their name or smth like that not just printing them out

File: client/chatclient/chat_client.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// creation timestamp, body-content and prints response from server.
+func (c *Client) SendMessage(m *Message) {
+	mes := &pb.Message{LoginFrom: m.LoginFrom, LoginTo: m.LoginTo, CreatedAt: m.CreatedAt, Body: m.Body}
+	ctx := context.Background()
+	r, err := c.client.SendMessage(ctx, &pb.SendMessageRequest{Message: mes})
+	if err != nil {
+		log.Fatalf("could not send message: %v", err)
+	}
+	log.Printf("Status: %s", r.Status)
+}
```
Comment (mmishche) [inline client/chatclient/chat_client.go:63]: the same here

File: client/chatclient/chat_client.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		mes, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Fatalf("Cannot receive: %v", err)
+		}
+		log.Printf("Message: %s", mes)
+	}
+}
```
Comment (mmishche) [inline client/chatclient/chat_client.go:82]: the same

File: client/main.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		client.GetUsers()
+	case "message":
+		created_at := int32(time.Now().Unix())
+		m := cl.Message{LoginFrom: *from, LoginTo: *to, CreatedAt: created_at, Body: *body}
+		client.SendMessage(&m)
+	case "subscribe":
+		client.Subscribe(*login)
+	case "":
+		log.Printf("Choose action")
+	}
```
Comment (mmishche) [inline client/main.go:39]: I'd rather move this piece of code to another method and call it a "submitRequest"

File: go.mod [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	google.golang.org/protobuf v1.28.0
+)
+
+require (
+	github.com/golang/protobuf v1.5.2 // indirect
+	golang.org/x/net v0.0.0-20200822124328-c89045814202 // indirect
+	golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd // indirect
+	golang.org/x/text v0.3.0 // indirect
+	google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013 // indirect
+)
```
[the remark this answers is not in the document]
Comment (mmishche) [reply go.mod:16]: It's okay, it means that packages are used indirect way

File: server/main.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// GetUsers returns list of users.
+func (s *server) GetUsers(ctx context.Context, in *pb.GetUsersRequest) (*pb.GetUsersReply, error) {
+	return &pb.GetUsersReply{Users: []*pb.User{{Login: "userA", FullName: "aa aa"},
+		{Login: "userB", FullName: "bb bb"}}}, nil
+}
+
+// SendMessage gets message and returns simple string if the message from client is received.
+func (s *server) SendMessage(ctx context.Context, in *pb.SendMessageRequest) (*pb.SendMessageReply, error) {
+	return &pb.SendMessageReply{Status: in.Message.LoginTo +
+		" received message from " + in.Message.LoginFrom}, nil
```
Comment (mmishche) [inline server/main.go:32]: To make a code a bit easier to understand, I would define a new var and call it "statusMessage" instead concatenate it during defining an instance

File: client/chatclient/chat_client.go (hunk 2 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go]: Thanks, got it. Changed.

File: client/chatclient/chat_client.go (hunk 3 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go]: Done.

File: client/chatclient/chat_client.go (hunk 4 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go]: Done.

File: client/main.go (hunk 5 above)
Comment (MetalMari) [reply client/main.go]: Thanks, done.

File: server/main.go (hunk 7 above)
Comment (MetalMari) [reply server/main.go]: Done

File: client/chatclient/chat_client.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+// GetUsers gets and prints list of users.
+func (c *Client) GetUsers() (Users []*pb.User, err error) {
+	ctx := context.Background()
+
+	r, err := c.client.GetUsers(ctx, &pb.GetUsersRequest{})
+	if err != nil {
+		log.Fatalf("could not get users: %v", err)
+	}
+	Users = r.Users
+	return Users, nil
```
Comment (mmishche) [inline client/chatclient/chat_client.go:51]: If you specify variables in the return state, you can write just return

File: client/chatclient/chat_client.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+// creation timestamp, body-content and prints response from server.
+func (c *Client) SendMessage(m *Message) (Status string, err error) {
+	mes := &pb.Message{LoginFrom: m.LoginFrom, LoginTo: m.LoginTo, CreatedAt: m.CreatedAt, Body: m.Body}
+	ctx := context.Background()
+	r, err := c.client.SendMessage(ctx, &pb.SendMessageRequest{Message: mes})
+	if err != nil {
+		log.Fatalf("could not send message: %v", err)
+	}
+	Status = r.Status
+	return Status, err
```
Comment (mmishche) [inline client/chatclient/chat_client.go:71]: the same as above

File: client/chatclient/chat_client.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		mes, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Fatalf("Cannot receive: %v", err)
+		}
+		Messages = append(Messages, mes)
+	}
+	return Messages, nil
```
Comment (mmishche) [inline client/chatclient/chat_client.go:91]: the same as above

**Review by mmishche — APPROVED** [review]

File: client/chatclient/chat_client.go (hunk 8 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go]: Thanks, done.

File: client/chatclient/chat_client.go (hunk 10 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go:89]: Done

File: client/chatclient/chat_client.go (hunk 9 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go]: Done

File: client/chatclient/chat_client.go (hunk 10 above)
Comment (MetalMari) [reply client/chatclient/chat_client.go]: Done.

**Pull request merged**